src/libraries/System.Private.CoreLib/src/System/Reflection/Emit/TypeNameBuilder.cs

```diff
@@ -0,0 +1,431 @@
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+//
+//  This TypeNameBuilder is ported from CoreCLR's original.
+//  It replaces the C++ bits of the implementation with a faithful C# port.
+
+using System.Collections.Generic;
+using System.Collections;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+using System.Text;
+using System.Diagnostics;
+
+namespace System.Reflection.Emit
+{
+    internal class TypeNameBuilder
+    {
+        private enum ParseState
+        {
+            Start = 0x0001,
+            Name = 0x0004,
+            GenArgs = 0x0008,
+            PtrArr = 0x0010,
+            ByRef = 0x0020,
+            AssemblySpec = 0x0080,
+        }
+
+        private ParseState _parseState;
+        private StringBuilder _str = new StringBuilder();
+        private int _instNesting;
+        private bool _firstInstArg;
+        private bool _nestedName;
+        private bool _hasAssemblySpec;
+        private bool _useAngleBracketsForGenerics;
+        private List<int> _stack = new List<int>();
+        private int _stackIdx;
+
+        public TypeNameBuilder()
+        {
+            _parseState = ParseState.Start;
+        }
+
+        public void OpenGenericArguments()
+        {
+            CheckParseState(ParseState.Name);
+
+            _parseState = ParseState.Start;
+            _instNesting++;
+            _firstInstArg = true;
+
+            if (_useAngleBracketsForGenerics)
+                Append('<');
+            else
+                Append('[');
+        }
+
+        public void CloseGenericArguments()
+        {
+            CheckParseState(ParseState.Start);
+
+            Debug.Assert(_instNesting != 0);
+
+            _parseState = ParseState.GenArgs;
+
+            _instNesting--;
+
+            if (_firstInstArg)
+            {
+                _str.Remove(_str.Length - 1, 1);
+            }
+            else
+            {
+                if (_useAngleBracketsForGenerics)
+                    Append('>');
+                else
+                    Append(']');
+            }
+        }
+
+        public void OpenGenericArgument()
+        {
+            CheckParseState(ParseState.Start);
+
+            Debug.Assert(_instNesting != 0);
+
+            _parseState = ParseState.Start;
+            _nestedName = false;
+
+            if (!_firstInstArg)
+                Append(',');
+
+            _firstInstArg = false;
+
+            if (_useAngleBracketsForGenerics)
+                Append('<');
+            else
+                Append('[');
+
+            PushOpenGenericArgument();
+        }
+
+        public void CloseGenericArgument()
+        {
+            CheckParseState(ParseState.Name | ParseState.GenArgs | ParseState.PtrArr | ParseState.ByRef | ParseState.AssemblySpec);
+
+            Debug.Assert(_instNesting != 0);
+
+            _parseState = ParseState.Start;
```

Comment:
If you would like to keep this, the different parser states need good comments about what they mean. For example, why we are switching to `Start` state here?

Also, I would rename ParserState to FormatterState. This type is not parsing anything.

Reply:
@jkotas I can't provide great comments since these are not mine - alright I will just delete them.